include/llzk/Util/ErrorHelper.h

```diff
@@ -16,19 +16,136 @@
 #include <llvm/ADT/Twine.h>
 #include <llvm/Support/ErrorHandling.h>
 
+#include <variant>
+
 namespace llzk {
 
-using EmitErrorFn = llvm::function_ref<mlir::InFlightDiagnostic()>;
+/// Wrapper around InFlightDiagnostic that can either be a regular InFlightDiagnostic or a
+/// special version that asserts false after reporting the diagnostic.
+/// See `wrapNullableInFlightDiagnostic()` below for details.
+class InFlightDiagnosticWrapper {
+private:
+  /// Implementation that gives an assertion failure after reporting.
+  /// See `wrapNullableInFlightDiagnostic()` below for details.
+  class DefaultAndFailInFlightDiagnostic : public mlir::InFlightDiagnostic {
+  public:
+    DefaultAndFailInFlightDiagnostic(mlir::InFlightDiagnostic &&base)
+        : mlir::InFlightDiagnostic(std::move(base)) {}
+
+    DefaultAndFailInFlightDiagnostic(DefaultAndFailInFlightDiagnostic &&other)
+        : mlir::InFlightDiagnostic(std::move(other)) {}
+
+    ~DefaultAndFailInFlightDiagnostic() {
+      if (getUnderlyingDiagnostic()) {
+        report();
+      }
+    }
+
+    void report() {
+      InFlightDiagnostic::report();
+      assert(false);
+    }
+  };
+
+  std::variant<mlir::InFlightDiagnostic, DefaultAndFailInFlightDiagnostic> inner;
+
+  explicit InFlightDiagnosticWrapper(DefaultAndFailInFlightDiagnostic &&diag)
+      : inner(std::move(diag)) {}
+
+public:
+  // Constructor for regular InFlightDiagnostic.
+  explicit InFlightDiagnosticWrapper(mlir::InFlightDiagnostic &&diag) : inner(std::move(diag)) {}
+
+  // Constructor for DefaultAndFailInFlightDiagnostic from MLIRContext.
+  // NOTE: This is not a common use case since it will always result in an assertion failure
+  // immediately after reporting the error; likely only useful in custom type builders.
+  explicit InFlightDiagnosticWrapper(mlir::MLIRContext *ctx)
+      : InFlightDiagnosticWrapper(
+            DefaultAndFailInFlightDiagnostic(mlir::detail::getDefaultDiagnosticEmitFn(ctx)())
+        ) {}
+
+  // Constructor for DefaultAndFailInFlightDiagnostic from Location.
+  // NOTE: This is not a common use case since it will always result in an assertion failure
+  // immediately after reporting the error; likely only useful in custom type builders.
```

Comment:
I think the constructor comments can be converted into doxygen comments (e.g., `///` vs `//`)